src/cursor/parse_cursor.rs

```diff
@@ -0,0 +1,567 @@
+//! Parse the contents of a cursor file
+
+// This code is loosely based on parse_cursor_file.c from libxcb-cursor, which is:
+//   Copyright © 2013 Michael Stapelberg
+// and is covered by MIT/X Consortium License
+
+use std::convert::TryInto;
+use std::io::{Read, Seek, SeekFrom};
+
+const FILE_MAGIC: u32 = 0x7275_6358;
+const IMAGE_TYPE: u32 = 0xfffd_0002;
+const IMAGE_MAX_SIZE: u32 = 0x7fff;
+
+/// An error that occurred while parsing
+#[derive(Debug)]
+pub(crate) enum Error {
+    /// An I/O error occurred
+    IO(std::io::Error),
+
+    /// The file did not begin with the expected magic
+    InvalidMagic,
+
+    /// The file contained unrealistically many images
+    TooManyEntries,
+
+    /// The file contains no images
+    NoImages,
+
+    /// Some image's entry is corrupt and not self-consistent
+    CorruptImage,
+
+    /// An entry is larger than the maximum size
+    ImageTooLarge,
+}
+
+impl From<std::io::Error> for Error {
+    fn from(value: std::io::Error) -> Self {
+        Error::IO(value)
+    }
+}
+
+/// Read a single `u32` from a cursor file
+///
+/// The file stores these entries in little endian.
+fn read_u32<R: Read>(read: &mut R) -> Result<u32, Error> {
+    let mut buffer = [0; 4];
+    read.read_exact(&mut buffer)?;
+    Ok(u32::from_le_bytes(buffer))
+}
+
+/// A single cursor image
+#[derive(Debug)]
+pub(crate) struct Image {
+    pub(crate) width: u16,
+    pub(crate) height: u16,
+    pub(crate) x_hot: u16,
+    pub(crate) y_hot: u16,
+    pub(crate) delay: u32,
+    pub(crate) pixels: Vec<u32>,
+}
+
+impl Image {
+    /// Read an `Image` from a reader
+    fn read<R: Read>(read: &mut R, expected_kind: u32, expected_size: u32) -> Result<Self, Error> {
+        let (_header, kind, size, _version) = (
+            read_u32(read)?,
+            read_u32(read)?,
+            read_u32(read)?,
+            read_u32(read)?,
+        );
+        if (kind, size) != (expected_kind, expected_size) {
+            return Err(Error::CorruptImage);
+        }
+
+        let (width, height) = (read_u32(read)?, read_u32(read)?);
+        let (x_hot, y_hot) = (read_u32(read)?, read_u32(read)?);
+        let delay = read_u32(read)?;
+
+        if width > IMAGE_MAX_SIZE || height > IMAGE_MAX_SIZE {
+            return Err(Error::ImageTooLarge);
+        }
+        let (width, height) = match (width.try_into(), height.try_into()) {
+            (Ok(w), Ok(h)) => (w, h),
+            _ => unreachable!("The check above makes sure u16 is enough"),
+        };
+        let (x_hot, y_hot) = match (x_hot.try_into(), y_hot.try_into()) {
+            (Ok(x), Ok(y)) => (x, y),
+            _ => return Err(Error::ImageTooLarge),
+        };
```

Comment:
Does this look better? For `width` and `height`, it gets rid of the `unreachable!`. Of course, requires making `IMAGE_MAX_SIZE` a `u16` instead of `u32`.

```suggestion
        let width = read_u32(read)?
            .try_into()
            .ok()
            .filter(|&width| width <= IMAGE_MAX_SIZE)
            .ok_or(Error::ImageTooLarge)?;
        let height = read_u32(read)?
            .try_into()
            .ok()
            .filter(|&height| height <= IMAGE_MAX_SIZE)
            .ok_or(Error::ImageTooLarge)?;
        let x_hot = read_u32(read)?.try_into().or(Err(Error::ImageTooLarge))?;
        let y_hot = read_u32(read)?.try_into().or(Err(Error::ImageTooLarge))?;
        let delay = read_u32(read)?;
```

Reply:
Thanks for the suggestion. I kept the `read_u32(read)` closer together so that it is easier to see in which order fields are read and moved the long conversion chain to a helper function so that it is only necessary once, but besides that, I went with your suggestion.

I am amazed that you can write correct code in a GitHub comment that involved `Result::or` and `Option::ok_or`. I always have to look these up.